ground/src/main/java/com/google/android/ground/converter/GeometryModelToLocalDbConverter.kt

```diff
@@ -0,0 +1,115 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.ground.converter
+
+import com.google.android.ground.model.geometry.*
+import com.google.android.ground.persistence.local.room.entity.GeometryEntity
+import com.google.android.ground.persistence.local.room.models.Coordinates
+import com.google.android.ground.persistence.local.room.models.GeometryType
+import com.google.android.ground.util.toImmutableList
+import com.google.common.collect.ImmutableList
+import com.google.common.reflect.TypeToken
+import com.google.gson.Gson
+
+/** Converts [Geometry] model objects to/from a persisted database representation. */
+object GeometryModelToLocalDbConverter : Converter<Geometry, GeometryEntity> {
+  override fun convertTo(model: Geometry): GeometryEntity? =
+    when (model) {
+      is Point -> fromPointModel(model)
+      is Polygon -> fromPolygonModel(model)
+      else -> null
+    }
+
+  override fun convertFrom(entity: GeometryEntity): Geometry? =
+    when (entity.geometryType) {
+      GeometryType.POINT.name -> toPointModel(entity)
+      GeometryType.POLYGON.name -> toPolygonModel(entity)
+      else -> null
+    }
+
+  private fun toPointModel(record: GeometryEntity): Geometry? = record.location?.toPoint()
+
+  private fun toPolygonModel(record: GeometryEntity): Geometry {
+    val shell = LinearRing(parseVertices(record.vertices).map { it.coordinate })
+    val holes = parseHoles(record.holes).map { LinearRing(it.map(Point::coordinate)) }
+
+    return Polygon(shell, holes)
+  }
+
+  private fun fromPointModel(point: Point): GeometryEntity =
+    GeometryEntity(GeometryType.POINT.name, Coordinates.fromPoint(point))
+
+  private fun fromPolygonModel(polygon: Polygon): GeometryEntity {
+    val shell = formatVertices(polygon.vertices)
+    val holes = formatHoles(polygon.holes.map { it.vertices })
+
+    return GeometryEntity(GeometryType.POLYGON.name, null, shell, holes)
+  }
+
+  private fun formatHoles(holes: List<ImmutableList<Point>>): String? {
+    if (holes.isEmpty()) {
+      return null
+    }
+
+    val gson = Gson()
+    val holeArray =
+      holes.map { hole -> hole.map { ImmutableList.of(it.coordinate.x, it.coordinate.y) } }
+
+    return gson.toJson(holeArray)
+  }
+
+  fun formatVertices(vertices: ImmutableList<Point>): String? {
+    if (vertices.isEmpty()) {
+      return null
+    }
+    val gson = Gson()
+    val verticesArray =
+      vertices.map { (coordinate): Point -> ImmutableList.of(coordinate.x, coordinate.y) }.toList()
+    return gson.toJson(verticesArray)
+  }
+
+  private fun parseHoles(holes: String?): List<ImmutableList<Point>> {
+    if (holes == null || holes.isEmpty()) {
```

Comment:
nit: Can be replaced with `.isNullOrEmpty()`

Reply:
done, thanks!